### ISSUE TITLE
Disallow access to instance variable via class

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#240]
Currently an instance variable can be accessed using the class, which is wrong. For example, this code is accepted by the type checker:

```
class A:
    def f(self) -> None:
        self.x = 1
A.x = 1 # accepted, but should be an error
```

Class variables need to be assigned in the class body:

```
class A:
    x = 1
A.x = 1 # ok, and should be ok
```

**Comment by JukkaL** [comment]
Actually, I'm no longer convinced that this is a bug. Turning this into an enhancement proposal.

**Comment by gvanrossum** [comment]
Yeah, this may just be a somewhat unconventional wy of setting the default.

**Comment by JukkaL** [comment]
Postponing this as I'm not sure if this is a good idea but might still be worth considering.

**Comment by gsakkis** [comment]
I ran into a situation which is almost the opposite problem of the original one: not only a given attribute is deliberately accessible by both the instance and the class, but the type in each case different. Unsurprisingly mypy complains about `Incompatible types in assignment`.

Minimal example:
```py
class A:
    x = "a"

    def __init__(self) -> None:
        self.x = 1

assert A().x == 1
assert A.x == "a"
```

I tried annotating `x` both as an instance and class variable but this caused a `Name "x" already defined` error without getting rid of the first one:
```py
class A:
    x: int
    x: ClassVar[str] = "a"

    def __init__(self) -> None:
        self.x = 1
```

**Comment by vtgn** [comment]
Hello!
2024 here! What are the news since 2013?!

With Python 3.12.6 and mypy 1.11.2, mypy says still anything on this case, but Pylint and Pylance do it, and there is an error during the code execution.

Plus, for this close similar case:
```Python
class A:
    id: Final[int] = 5

    def __init__(self):
        self.id = 10

a = A()

print(A.id) # => 5
print(a.id) # => 10
print(a.__class__.id) # => 5
```
mypy doesn't say anything, which is correct because self.id defined in the constructor is not the same attribute as A.id, but I don't know if mypy says nothing because it manages correctly this case or it doesn't check anything. On the other hand, Pylance notifies _self.id_ in the constructor can't be overwritten because _A.id_ is Final, which is a bug because for Python, assignment on self.attr does not reference a class attribute, but an instance one! >_<...
